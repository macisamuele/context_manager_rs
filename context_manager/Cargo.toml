[package]
name = "context_manager"
version = "0.1.2"
edition = "2021"
description = "Python's like context_managers in Rust"
license = "MIT OR Apache-2.0"
repository = "https://github.com/macisamuele/context_manager_rs"
documentation = "https://docs.rs/context_manager_macro"
readme = "README.md"

[dependencies]
context_manager_macro = "=0.1.2"

[dev-dependencies]
tokio = {version = "1", features = ["macros", "rt", "rt-multi-thread"] }
trybuild = "1"
