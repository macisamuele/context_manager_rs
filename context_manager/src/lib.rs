#![deny(
    dead_code,
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms,
    rustdoc::all,
    rustdoc::missing_crate_level_docs,
    unreachable_pub,
    unused_imports,
    unused_variables
)]
// Use README.md file as module documentation
// This makes easy to have a proper home for the github project as well as
// ensuring that the content is always up to date and tested
#![doc = include_str!("../README.md")]
//!
#![doc = include_str!("../CHANGELOG.md")]

mod t_async;
mod t_sync;
pub use crate::t_async::AsyncWrapContext;
pub use crate::t_sync::SyncWrapContext;

/// Context about the caller propagated into the context.
#[derive(Debug)]
#[non_exhaustive]
pub struct CallerContext {
    /// Name of the wrapped function
    fn_name: &'static str,
}

impl CallerContext {
    /// Create a new instance of the `CallerContext`
    #[must_use]
    pub fn new(fn_name: &'static str) -> Self {
        Self { fn_name }
    }

    /// Name of the wrapped function
    #[must_use]
    pub fn fn_name(&self) -> &str {
        self.fn_name
    }
}

/// Procedural macro that will decorate the incoming async function with the provided context.
///
/// The context is expected to be a type that implements the `AsyncWrapContext` trait.
///
/// Usage example:
/// ```
/// # use context_manager_macro::async_wrap;
/// struct AsyncPrintDuration;
/// impl<T> context_manager::AsyncWrapContext<T> for AsyncPrintDuration {
///   async fn new() -> Self { Self }
/// }
///
/// #[async_wrap(AsyncPrintDuration)]
/// async fn foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     let type_name = std::any::type_name::<T>();
///     println!("Async call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///     10
/// }
/// ```
///
/// The decorator does not induce limits on the shape of the incoming function, in terms
/// of generics, sync/async, lifetime, etc.
///
/// The decorator will expand the incoming function by adding the context handling
/// rendering something similar to
/// ```
/// # use context_manager::{AsyncWrapContext, CallerContext};
/// # struct AsyncPrintDuration;
/// # impl<T> AsyncWrapContext<T> for AsyncPrintDuration {
/// #   async fn new() -> Self { Self }
/// # }
/// async fn foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     AsyncPrintDuration::run(CallerContext { fn_name: "foo" }, async {
///         let type_name = std::any::type_name::<T>();
///         println!("Async call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///         10
///     }).await
/// }
/// ```
///
/// The structuring of the generated code is though to avoid any clone/copy of data,
/// as well as reducing the number of jumps needed to execute the original code.
///
/// # Possible compile errors
/// ## Passing a type that does not implement `AsyncWrapContext` trait will lead to compile errors.
/// ```compile_fail
/// # use context_manager_macro::async_wrap;
/// struct PrintDuration;
/// impl<T> context_manager::SyncWrapContext<T> for PrintDuration {
///   fn new() -> Self { Self }
/// }
///
/// #[async_wrap(PrintDuration)]
/// async fn foo() {}
/// ```
/// would lead to the following compile error
/// ```text
/// error[E0277]: the trait bound `PrintDuration: AsyncWrapContext<_>` is not satisfied
///   --> src/lib.rs:206:1
///    |
/// 11 | #[async_wrap(PrintDuration)]
///    |                  ^^^^^^^^^^^^^ the trait `AsyncWrapContext<_>` is not implemented for `PrintDuration`
///    |
///    = help: the trait `AsyncWrapContext<T>` is implemented for `AsyncTraceDuration`
/// ```
///
/// ## Decorating a synchronous function
/// This is intentional in order to avoid possibly stalling the async-runtime in use.
/// Please consider wrapping yourself the synchronous code in an async block, or using `#[decorate]` whether possible.
/// ```compile_fail
/// # use context_manager_macro::async_wrap;
/// struct AsyncPrintDuration;
/// impl<T> context_manager::AsyncWrapContext<T> for AsyncPrintDuration {
///   async fn new() -> Self { Self }
/// }
///
/// #[async_wrap(PrintDuration)]
/// fn foo() {}
/// ```
/// would lead to the following error
/// ```text
/// error: #[async_wrap] cannot operate on sync functions. Please consider using a #[decorate] macro or converting/wrapping the function to be async.
///   --> src/lib.rs:228:1
///    |
/// 11 | #[async_wrap(PrintDuration)]
///    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
///    |
/// ```
pub use context_manager_macro::async_wrap;

/// Procedural macro that will decorate the incoming function with the provided context.
///
/// The context is expected to be a type that implements the `SyncWrapContext` trait.
///
/// Usage example:
/// ```
/// # use context_manager_macro::wrap;
/// struct PrintDuration;
/// impl<T> context_manager::SyncWrapContext<T> for PrintDuration {
///   fn new() -> Self { Self }
/// }
///
/// #[wrap(PrintDuration)]
/// fn sync_foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     let type_name = std::any::type_name::<T>();
///     println!("Sync call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///     10
/// }
///
/// #[wrap(PrintDuration)]
/// async fn async_foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     let type_name = std::any::type_name::<T>();
///     println!("Async call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///     10
/// }
/// ```
///
/// The decorator does not induce limits on the shape of the incoming function, in terms
/// of generics, sync/async, lifetime, etc.
///
/// The decorator will expand the incoming function by adding the context handling
/// rendering something similar to
/// ```
/// # use context_manager::{CallerContext, SyncWrapContext};
/// # struct PrintDuration;
/// # impl<T> SyncWrapContext<T> for PrintDuration {
/// #   fn new() -> Self { Self }
/// # }
/// fn sync_foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     PrintDuration::run_sync(CallerContext { fn_name: "sync_foo" }, move || {
///         let type_name = std::any::type_name::<T>();
///         println!("Sync call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///         10
///     })
/// }
///
/// async fn async_foo<'a, T>(int_value: usize, str_ref: &'a str, generic: T) -> usize {
///     PrintDuration::run_async(CallerContext { fn_name: "async_foo" }, async {
///         let type_name = std::any::type_name::<T>();
///         println!("Async call with int_value={int_value}, str_ref={str_ref}, type_of(T)={type_name}");
///         10
///     }).await
/// }
/// ```
///
/// The structuring of the generated code is though to avoid any clone/copy of data,
/// as well as reducing the number of jumps needed to execute the original code.
///
/// # Possible compile errors
/// ## Passing a type that does not implement `SyncWrapContext` trait will lead to compile errors.
/// ```compile_fail
/// # use context_manager_macro::wrap;
/// struct AsyncPrintDuration;
/// impl<T> context_manager::AsyncWrapContext<T> for AsyncPrintDuration {
///   async fn new() -> Self { Self }
/// }
///
/// #[wrap(AsyncPrintDuration)]
/// fn foo() {}
/// ```
/// would lead to the following compile error
/// ```text
/// ---- src/lib.rs - decorate (line 98) stdout ----
/// error[E0277]: the trait bound `AsyncPrintDuration: SyncWrapContext<_>` is not satisfied
///   --> src/lib.rs:106:12
///    |
/// 11 | #[wrap(AsyncPrintDuration)]
///    |            ^^^^^^^^^^^^^^^^^^ the trait `SyncWrapContext<_>` is not implemented for `AsyncPrintDuration`
///    |
///    = help: the trait `SyncWrapContext<T>` is implemented for `TraceDuration`
/// ```
///
/// ## Decorating a constant function
/// Const functions are not supported for decoration.
/// This is a side-effect of embedding code that is not const compatible (like async blocks and closures).
///
/// ```compile_fail
/// # use context_manager_macro::wrap;
/// struct PrintDuration;
/// impl<T> context_manager::SyncWrapContext<T> for PrintDuration {
///   fn new() -> Self { Self }
/// }
///
/// #[wrap(PrintDuration)]
/// const fn foo() {}
/// ```
/// would lead to the following error
/// ```text
/// error: #[wrap] cannot operate on const functions
///   --> context_manager_macro/src/lib.rs:131:1
///    |
/// 11 | #[wrap(PrintDuration)]
///    | ^^^^^^^^^^^^^^^^^^^^^^
///    |
/// ```
pub use context_manager_macro::wrap;

#[cfg(test)]
mod tests {
    use trybuild::TestCases;

    #[test]
    fn procedural_macros_ui_tests() {
        let t = TestCases::new();
        t.pass("tests/ui/pass/*.rs");
        t.compile_fail("tests/ui/fail/*.rs");
    }
}
