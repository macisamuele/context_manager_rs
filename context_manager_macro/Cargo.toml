[package]
name = "context_manager_macro"
version = "0.1.0"
edition = "2021"
description = "Process macro for context_manager crate"
license = "MIT OR Apache-2.0"
repository = "https://github.com/macisamuele/context_manager_rs"
documentation = "https://docs.rs/context_manager_macro"
readme = "../README.md"

[lib]
proc-macro = true

[dependencies]
syn = {version = "2", features = ["full"] }
quote = "1"
