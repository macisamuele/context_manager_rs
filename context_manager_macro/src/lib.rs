#![deny(
    dead_code,
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms,
    rustdoc::all,
    rustdoc::missing_crate_level_docs,
    unreachable_pub,
    unused_imports,
    unused_variables
)]

//! Implementation of the procedural macros exposed by [`context_manager`](https://crates.io/crates/context-manager) crate.

use proc_macro::TokenStream;
use quote::quote;
use syn::parse::Parse;
use syn::parse::ParseStream;
use syn::parse_macro_input;
use syn::parse_quote;
use syn::Block;
use syn::Error;
use syn::ItemFn;
use syn::Type;

struct Args {
    context_type: Type,
}

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        if input.is_empty() {
            Err(Error::new(
                input.span(),
                "Expected a type as argument: `#[wrap(Type)]` or `#[async_wrap(Type)]`",
            ))
        } else {
            Ok(Self {
                context_type: input.parse::<Type>()?,
            })
        }
    }
}

/// Procedural macro that will decorate the incoming function with the provided context.
///
/// The context is expected to be a type that implements the `context_manager::SyncWrapContext` trait.
///
/// More documentation available [here](https://docs.rs/context_manager/latest/context_manager/attr.wrap.html)
#[proc_macro_attribute]
pub fn wrap(attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut in_func = parse_macro_input!(item as ItemFn);

    if in_func.sig.constness.is_some() {
        // Insert compile error at the begin of the function block.
        // Doing so allows a clear compile failure, while allowing type inference to still work.
        in_func.block.stmts.insert(
            0,
            parse_quote!(::std::compile_error!("#[wrap] cannot operate on const functions.");),
        );
        return quote! { #in_func }.into();
    };

    let args: Args = parse_macro_input!(attr);

    let context_type = &args.context_type;
    let block = &in_func.block;
    let new_body: TokenStream = if in_func.sig.asyncness.is_some() {
        quote! {
            {
                <#context_type as ::context_manager::SyncWrapContext<_>>::run_async(async #block).await
            }
        }
        .into()
    } else {
        quote! {
            {
                <#context_type as ::context_manager::SyncWrapContext<_>>::run_sync(move || #block)
            }
        }
        .into()
    };

    in_func.block.stmts = parse_macro_input!(new_body as Block).stmts;

    quote! { #in_func }.into()
}

/// Procedural macro that will decorate the incoming async function with the provided context.
///
/// The context is expected to be a type that implements the `context_manager::AsyncWrapContext` trait.
///
/// More documentation available [here](https://docs.rs/context_manager/latest/context_manager/attr.async_wrap.html)
#[proc_macro_attribute]
pub fn async_wrap(attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut in_func = parse_macro_input!(item as ItemFn);

    if in_func.sig.constness.is_some() {
        // This is not really possible, because "functions cannot be both `const` and `async`"
        // but let's keep this check for future-proofing
        // Insert compile error at the begin of the function block.
        // Doing so allows a clear compile failure, while allowing type inference to still work.
        in_func.block.stmts.insert(
            0,
            parse_quote!(::std::compile_error!("#[wrap] cannot operate on const functions.");),
        );
        return quote! { #in_func }.into();
    };

    let args: Args = parse_macro_input!(attr);

    let context_type = &args.context_type;
    let block = &in_func.block;
    if in_func.sig.asyncness.is_some() {
        let new_body: TokenStream = quote! {
            {
                <#context_type as ::context_manager::AsyncWrapContext<_>>::run(async #block).await
            }
        }
        .into();
        in_func.block.stmts = parse_macro_input!(new_body as Block).stmts;
    } else {
        // Insert compile error at the begin of the function block.
        // Doing so allows a clear compile failure, while allowing type inference to still work.
        in_func.block.stmts.insert(
            0,
            parse_quote!({::std::compile_error!(
                "#[async_wrap] cannot operate on sync functions. Please consider using a #[wrap] macro or converting/wrapping the function to be async."
            )})
        );
    };

    quote! { #in_func }.into()
}
