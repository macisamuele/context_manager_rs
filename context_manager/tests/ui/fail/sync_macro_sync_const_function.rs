use context_manager::SyncWrapContext;
use context_manager_macro::wrap;

struct Sync;
impl<T> SyncWrapContext<T> for Sync {
    fn new() -> Self {
        Self
    }
}

#[wrap(Sync)]
const fn sync_foo<'a, T: ?Sized>(v: &'a T) -> &'a T {
    v
}

fn main() {
    assert_eq!(sync_foo("10"), "10");
}
