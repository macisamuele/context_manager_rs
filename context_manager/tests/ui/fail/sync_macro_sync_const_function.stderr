error: #[wrap] cannot operate on const functions.
  --> tests/ui/fail/sync_macro_sync_const_function.rs:11:1
   |
11 | #[wrap(Sync)]
   | ^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `wrap` (in Nightly builds, run with -Z macro-backtrace for more info)
